src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/LambdaExpression.cs

```diff
@@ -154,7 +154,7 @@ public Delegate Compile()
         /// <returns>A delegate containing the compiled version of the lambda.</returns>
         public Delegate Compile(bool preferInterpretation)
         {
-            if (CanCompileToIL && CanInterpret && preferInterpretation)
+            if (CanInterpret && preferInterpretation)
```

Comment:
Can you check if this impacts size with NativeAOT?

The dead branch elimination in NativeAOT only considers things that have feature switches. It doesn't attempt to inline things.

I think this is changing things from `if (false && CanInterpret && preferInterpretation)` to `if (CanInterpret && preferInterpretation)` (no longer a known constant).

Could be fixed by creating a substitution rule for CanInterpret as well.

(Either that or beefing up the dead branch elimination in NativeAOT. The dead branch elimination in NativeAOT is simpler than in linker because: 1. linker's one is super scary and I don't want to have to deal with the bugs in it that I would get by porting it, and 2. longer term, we should have RyuJIT do the dead branch elimination. RyuJIT is much better at it that anything we would write specifically for trimming.)

Reply:
Or maybe this is fine now that I look at it for longer?

Reply:
Yes, I think it makes sense to do. I'm responsible for introducing this in https://github.com/dotnet/runtime/pull/61952.

That pull request was a mechanical replacement of `#defines`. The code originally looked like:

```csharp
        public Delegate Compile(bool preferInterpretation)
        {
#if FEATURE_COMPILE && FEATURE_INTERPRET
            if (preferInterpretation)
            {
                return new Interpreter.LightCompiler().CompileTop(this).CreateDelegate();
            }
#endif
            return Compile();
        }
```

Reply:
On a second look, the original code doesn't make that much more sense either...